moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,572 @@
+    server_event["event_type"] = copy.copy(event["event_type"])
+    server_event["generated"] = copy.copy(event["event_time"])
+    return server_event
+
+
+def get_session_for_resource(resource):
+    auth_url = resource.get_attribute(attributes.RESOURCE_AUTH_URL)
+    # Note: Authentication for a specific OpenStack cloud is stored in env
+    # variables of the form OPENSTACK_{RESOURCE_NAME}_APPLICATION_CREDENTIAL_ID
+    # and OPENSTACK_{RESOURCE_NAME}_APPLICATION_CREDENTIAL_SECRET
+    # where resource name is has spaces replaced with underscored and is
+    # uppercase.
+    # This allows for the possibility of managing multiple OpenStack clouds
+    # via multiple resources.
+    var_name = resource.name.replace(" ", "_").replace("-", "_").upper()
+    auth = v3.ApplicationCredential(
+        auth_url=auth_url,
+        application_credential_id=os.environ.get(f"OPENSTACK_{var_name}_APPLICATION_CREDENTIAL_ID"),
+        application_credential_secret=os.environ.get(f"OPENSTACK_{var_name}_APPLICATION_CREDENTIAL_SECRET"),
+    )
+    return session.Session(auth, verify=os.environ.get("FUNCTIONAL_TESTS", "") != "True")
+
+
+def main():
+    """The main function"""
+    config = {}
+
+    do_parse_args(config)
+    do_read_config(config)
+
+    openstack_conn = openstack.connect(cloud=config["cloud"])
+    openstack_nova = nova_client.Client(2, session=openstack_conn.session)
+    openstack_cinder = cinder_client.Client(3, session=openstack_conn.session)
+    openstack_keystone = keystone_client.Client(3, session=openstack_conn.session)
+
+    script_timestamp = config["end"]
+    date_format_str = "%Y-%m-%dT%H:%M:%S"
+    last_run_timestamp = (datetime.datetime.strptime(script_timestamp, date_format_str) - datetime.timedelta(hours=1)).strftime(date_format_str)
+    vm_timestamps = {}
+    try:
+        with open("vm_last_report_time.json", "r", encoding="utf-8") as file:
+            reporting_state = json.load(file)
+        vm_timestamps = reporting_state.vm_timestamps
+        last_run_timestamp = reporting_state.last_run_timestamp
+    except IOError:
+        pass
+    except ValueError:
+        # Ignore badly formatted file or empty file
+        pass
+
+    # so that we can tell if the VM not present
+    for vm in vm_timestamps.values():
+        vm["updated"] = 0
+
+    flavor_dict = {}
+    for flavor in openstack_conn.list_flavors():
+        flavor_dict[flavor.id] = flavor
+
+    user_dict = {}
+    for user in openstack_conn.list_users():
+        user_dict[user.id] = user
+
+    project_dict = {}
+    for project in openstack_conn.list_projects():
+        project_dict[project.id] = project
+
+    volume_dict = {}
+    for volume in openstack_cinder.volumes.list(search_opts={"all_tenants": True}, detailed=True):
+        volume_dict[volume.id] = volume
+
+    events = []
+    server_dict = {}
+    ap_t = datetime.datetime.strptime(script_timestamp, date_format_str)
```

Comment:
what's `ap_t`?

Reply:
audit_period_time - honestly it isn't actually the audit_period_time, but the first timestamp in the series of events.  technically, the audit period shouldn't be based on the first timestamp in the series of events, it should be based on either an audit period for a grant, or an audit period for xdmod.

Have since found that xdmod is not using audit period for anything yet.  I have found series of events in the reference file that all events for a vm fall within a single audit period, and sometimes they are split into 2 audit periods.  So I can get away with setting it for the minimum event time to the current time of the script.  Note, that this is only done when the script runs for the first time.

Reply:
okay, I think having a more descriptive name would be better.

Reply:
So, how to put this, I am trying to get this to produce data.  I have renamed ap_t to something more meaningful.  In doing so, realized that this does need to be refactored to be easier to read.

So I am trying to balance the ease of reading with the urgency to produce data.